core/lib/zksync_core/src/consensus/mod.rs

```diff
@@ -61,61 +60,155 @@ impl MainNodeConfig {
     }
 }
 
-/// Periodically fetches the head of the main node
-/// and updates `SyncState` accordingly.
-pub async fn run_main_node_state_fetcher(
-    ctx: &ctx::Ctx,
-    client: &dyn MainNodeClient,
-    sync_state: &SyncState,
-) -> ctx::OrCanceled<()> {
-    const DELAY_INTERVAL: time::Duration = time::Duration::milliseconds(500);
-    const RETRY_DELAY_INTERVAL: time::Duration = time::Duration::seconds(5);
-    loop {
-        match ctx.wait(client.fetch_l2_block_number()).await? {
-            Ok(head) => {
-                sync_state.set_main_node_block(head);
-                ctx.sleep(DELAY_INTERVAL).await?;
+pub type FetcherConfig = executor::Config;
+
+/// External node consensus config.
+#[derive(Debug)]
+pub struct Fetcher {
+    pub config: FetcherConfig,
+    pub sync_state: SyncState,
+    pub client: Box<dyn MainNodeClient>,
+}
+
+impl Fetcher {
+    /// Periodically fetches the head of the main node
+    /// and updates `SyncState` accordingly.
+    async fn fetch_state_loop(&self, ctx: &ctx::Ctx) -> ctx::Result<()> {
+        const DELAY_INTERVAL: time::Duration = time::Duration::milliseconds(500);
+        const RETRY_INTERVAL: time::Duration = time::Duration::seconds(5);
+        loop {
+            match ctx.wait(self.client.fetch_l2_block_number()).await? {
+                Ok(head) => {
+                    self.sync_state.set_main_node_block(head);
+                    ctx.sleep(DELAY_INTERVAL).await?;
+                }
+                Err(err) => {
+                    tracing::warn!("main_node_client.fetch_l2_block_number(): {err}");
+                    ctx.sleep(RETRY_INTERVAL).await?;
+                }
             }
-            Err(err) => {
-                tracing::warn!("main_node_client.fetch_l2_block_number(): {err}");
-                ctx.sleep(RETRY_DELAY_INTERVAL).await?;
+        }
+    }
+
+    /// Fetches genesis from the main node.
+    async fn fetch_genesis(&self, ctx: &ctx::Ctx) -> ctx::Result<validator::Genesis> {
+        let genesis = ctx
+            .wait(self.client.fetch_consensus_genesis())
+            .await?
+            .context("fetch_consensus_genesis()")?
+            .context("main node is not running consensus component")?;
+        Ok(zksync_protobuf::serde::deserialize(&genesis.0).context("deserialize(genesis)")?)
+    }
+
+    /// Fetches (with retries) the given block from the main node.
+    async fn fetch_block(&self, ctx: &ctx::Ctx, n: MiniblockNumber) -> ctx::Result<FetchedBlock> {
+        const RETRY_INTERVAL: time::Duration = time::Duration::seconds(5);
+        loop {
+            let res = ctx.wait(self.client.fetch_l2_block(n, true)).await?;
+            match res {
+                Ok(Some(block)) => return Ok(block.try_into()?),
+                Ok(None) => {}
```

Comment:
Just to clarify: Is not having a block intentionally considered an error for the purpose of retries?

Reply:
It is definitely an error. The question is whether it should be considered transient or fatal. Main node is expected to have all the blocks before the genesis, so we can consider it fatal. It may happen that None will be received if the main node forks while we are syncing, but that will be detected by the genesis change detector, so here it is ok for it to be considered transient.